synapse/handlers/account_validity.py

```diff
@@ -70,6 +80,96 @@ def __init__(self, hs: "HomeServer"):
             if hs.config.run_background_tasks:
                 self.clock.looping_call(self._send_renewal_emails, 30 * 60 * 1000)
 
+        self._is_user_expired_callbacks: List[IS_USER_EXPIRED_CALLBACK] = []
+        self._on_user_registration_callbacks: List[ON_USER_REGISTRATION_CALLBACK] = []
+        self._on_legacy_send_mail_callback: Optional[
+            ON_LEGACY_SEND_MAIL_CALLBACK
+        ] = None
+        self._on_legacy_renew_callback: Optional[ON_LEGACY_RENEW_CALLBACK] = None
+
+        # The legacy admin requests callback isn't a protected attribute because we need
+        # to access it from the admin servlet, which is outside of this handler.
+        self.on_legacy_admin_request_callback: Optional[ON_LEGACY_ADMIN_REQUEST] = None
+
+    def register_account_validity_callbacks(
+        self,
+        is_user_expired: Optional[IS_USER_EXPIRED_CALLBACK] = None,
+        on_user_registration: Optional[ON_USER_REGISTRATION_CALLBACK] = None,
+        on_legacy_send_mail: Optional[ON_LEGACY_SEND_MAIL_CALLBACK] = None,
+        on_legacy_renew: Optional[ON_LEGACY_RENEW_CALLBACK] = None,
+        on_legacy_admin_request: Optional[ON_LEGACY_ADMIN_REQUEST] = None,
+    ):
+        """Register callbacks from module for each hook."""
+        if is_user_expired is not None:
+            self._is_user_expired_callbacks.append(is_user_expired)
+
+        if on_user_registration is not None:
+            self._on_user_registration_callbacks.append(on_user_registration)
+
+        # The builtin account validity feature exposes 3 endpoints (send_mail, renew, and
+        # an admin one). As part of moving the feature into a module, we need to change
+        # the path from /_matrix/client/unstable/account_validity/... to
+        # /_synapse/client/account_validity, because:
+        #   * the feature isn't part of the Matrix spec thus shouldn't live under /_matrix
+        #   * because of the way we register servlets modules can't register resources
+        #     under /_matrix/client
+        # We need to allow for a transition period between the old and new endpoints
+        # in order to allow for clients to update (and for emails to be processed).
+        # Therefore we need to allow modules (in practice just the one implementing the
+        # email-based account validity) to temporarily hook into the legacy endpoint so we
+        # can route the traffic coming into the old endpoints into the module, which is
+        # why we have the following three temporary hooks.
```

Comment:
given that we're keeping the old code around for now anyway, do we really need all of these hooks? For instance, can't we have the core code handle requests to `/_synapse/admin/v1/account_validity/validity` and the module handle requests to `/_synapse/client/account_validity/validity`, thus making the `on_legacy_admin_request` hook redundant?

Reply:
I think it would be confusing from a user pov because e.g. if they receive an email but click on the link 1w later (e.g. because they were on holidays) and the switch from the core code to the module happens in between, the link wouldn't be doing anything (since it would only update the old database table). The only other way to mitigate this that I can think of would be to have both the core code and the module enabled at the same time, but that'd mean users would get 2 emails with 2 different links to click on at about the same time, which is probably quite confusing.

Reply:
I'm not following you. We can handle `/_synapse/admin/v1/account_validity/validity` in the core while still having emails link to `/_synapse/client/account_validity/validity` ?

Reply:
We can, but then we'd end up with two different tables (`account_validity`, and `email_account_validity` from https://github.com/matrix-org/synapse-email-account-validity) used for the same purpose but with divergent data.

Reply:
ugh, I hadn't realised the new module was using a different table. Can it not use the same table?

Reply:
I'm a bit wary of having an external module write to a table that belongs to Synapse's schema to be honest. Especially since it's tied to a feature we'll be getting rid of eventually; having an external module use it means we would need extra care deleting the feature and its storage component and imho it's very easy to just forget it and only remember when it breaks e.g. on a customer's infrastructure.

Plus the naming wouldn't be right, since it then wouldn't just be account validity, but rather one way of doing account validity (the email-based one), and other modules might implement other ways to verify the validity of an account (e.g. by checking with an auth backend).

Reply:
hmmm yes, fair enough.